dnf/cli/commands/repoquery.py

```diff
@@ -343,55 +349,60 @@ def build_format_fn(self, opts, pkg):
-                                                      recursive=recursive, all_deps=all_deps)
-        return t.union(done)
+    def _resolve_nevras(self, nevras, base_query):
+        resolved_nevras = self.base.sack.query().filterm(empty=True)
+        for nevra in nevras:
+            resolved_nevras = resolved_nevras.union(base_query.intersection(
+                dnf.subject.Subject(nevra).get_best_query(
+                    self.base.sack,
+                    with_provides=False,
+                    with_filenames=False
+                )
+            ))
+
+        return resolved_nevras
+
+    def _do_recursive_deps(self, query_in, query_select, done=None):
+        done = done if done else query_select
+
+        query_required = query_in.filter(requires=query_select)
+
+        query_select = query_required.difference(done)
+        done = query_required.union(done)
+
+        if query_select:
+            done = self._do_recursive_deps(query_in, query_select, done=done)
 
-    def by_all_deps(self, requires_name, depends_name, query):
-        names = requires_name or depends_name
-        defaultquery = self.base.sack.query().filterm(empty=True)
-        for name in names:
-            defaultquery = defaultquery.union(query.intersection(
-                dnf.subject.Subject(name).get_best_query(self.base.sack, with_provides=False,
-                                                         with_filenames=False)))
-        requiresquery = query.filter(requires__glob=names)
-        if depends_name:
-            requiresquery = requiresquery.union(query.filter(recommends__glob=depends_name))
-            requiresquery = requiresquery.union(query.filter(enhances__glob=depends_name))
-            requiresquery = requiresquery.union(query.filter(supplements__glob=depends_name))
-            requiresquery = requiresquery.union(query.filter(suggests__glob=depends_name))
-
-        done = requiresquery.union(self._get_recursive_deps_query(query, defaultquery,
-                                                                  all_deps=depends_name))
-        if self.opts.recursive:
-            done = done.union(self._get_recursive_deps_query(query, done,
-                                                             recursive=self.opts.recursive,
-                                                             all_deps=depends_name))
         return done
 
+    def by_all_deps(self, names, query, all_dep_types=False):
+        # in case of arguments being NEVRAs, resolve them to packages
+        resolved_nevras = self._resolve_nevras(names, query)
+
+        # filter the arguments directly as reldeps
+        depquery = query.filter(requires__glob=names)
+
+        # filter the resolved NEVRAs as packages
+        depquery = depquery.union(query.filter(requires=resolved_nevras))
+
+        if all_dep_types:
+            # TODO this is actually very inefficient, as it builds the same list of
```

Comment:
Not sure about the comment. I would recommend to remove it because it has not much informative value.

Reply:
The comment has a lot of value, it just may not be phrased well. The problem is each call to query.filter IIRC resolves the `names` to packages or does some similarly expensive operation (sorry, I forgot the details, I can check them if you find it important here), which is the same for all the five total calls that are happening here. Therefore, if the code was written in a way this same operation was done only once, the code could be five times faster.

Reply:
Please could you remove the comment anyway?

Reply:
You need to give me a good reason for that.

Reply:
Simple I don't think that the code should be used as a wish list. Additionally I don't think that there is any real impact on performance. Also the implementation is limited by API available in libsolv.

Reply:
This is not a wish, it is a performance issue with the code that I've discovered.

There is a real impact on performance. I've done a very rudimentary performance check (as I kind of hope it will be enough for you) and for a repoquery like:

```
dnf3 repoquery --whatdepends "libdnf*"
```

The `pyseq_to_reldeplist()` function (https://github.com/rpm-software-management/libdnf/blob/20ba333f9a54e73e95257e1a78f0f55bb6e88809/python/hawkey/iutil-py.cpp#L237) is being called 5 times with the same arguments and the same result. It is taking about 50% of the total time (I've eyeballed the time tbh., didn't go through measuring it exactly...) of the command run time.

This is obviously not a libsolv limitation, it is a shortcoming of the current API design and the comment is there to bring attention to it, because it is not apparent at first glance this may be happening.

Reply:
Then please enhance the comment and provide additional information or remove it. The issue is with resolving string as a glob to reldep. It could be enhance by providing API for the conversion and then using list of reldeps for query. To general notes have nearly no value and the author of the note will not remember details. Just remember how many time you already mentioned that you do not remember details about your code, in this PR.

Reply:
I've adjusted the comment, sorry about being imprecise in the first place.

FTR I think the right API fix is to have support for `depends`, instead of separating out the resolving of globs. I think that's better for a high-level API.